#### Python code for Data analysis of the Granular dataset
#### By Ameya Ghatpande on Jan-18-2018

In [90]:
# <aghatpande> on 18-Jan-2018
# print all the outputs in a cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [91]:
# <aghatpande> on 18-Jan-2018
# Import all the libraries
import pandas as pd
import matplotlib.pyplot as plt

In [92]:
toyFieldsDF = pd.read_csv("toy_fields.csv", index_col=False)
print("Length of toys_fields is ", len(toyFieldsDF))
toyOrgsDF = pd.read_csv("toy_orgs.csv")
print("Length of toys_orgs is ", len(toyOrgsDF))
toyYieldsDF = pd.read_csv("toy_yields.csv")
print("Length of toys_yields is ", len(toyYieldsDF))

('Length of toys_fields is ', 417)
('Length of toys_orgs is ', 21)
('Length of toys_yields is ', 417)


### Lets try to solve the first expectation - We expect yield data to have one yield per org, year, field 

#### There is no Primary key associated with any of the tables. I am checking some sample values to see if there is 1:1 mapping between the toy_fields and toy_yields table

In [93]:
toyFieldsDF.loc[0:2, :]
toyYieldsDF.loc[0:2, :]

field_id  \
0  cb1e3bd821dd673a0c378b16ba97075919aa993a   
1  58d04a7094c1b08e9856932a02104d2a36b7b8ce   
2  5e3babe7b318f4245ceb16718665af0b54335443   

                                     org_id  year  field_area field_area_unit  
0  0966e34316e39217e4f26fbe20fc14e24c26567b  2016  307.185719            acre  
1  726cce9d9800d792d1e0fbb83923ac4c6be4f668  2014  213.680671            acre  
2  598d17ec200eeb9e27ae829746ccf47c57761bc1  2015  160.063342            acre

crop_type  year harvest_product  total_harvested harvested_unit  \
0  CROP.CORN.COMMERCIAL  2016           grain     53290.628852             bu   
1  CROP.CORN.COMMERCIAL  2014           grain     40073.821554             bu   
2  CROP.CORN.COMMERCIAL  2015           grain     45528.033941             bu   

   planted_area  harvested_area                                  field_id  \
0         40.38           560.0  cb1e3bd821dd673a0c378b16ba97075919aa993a   
1           NaN             NaN  58d04a7094c1b08e9856932a02104d2a36b7b8ce   
2        148.61             NaN  5e3babe7b318f4245ceb16718665af0b54335443   

                                     org_id  
0  0966e34316e39217e4f26fbe20fc14e24c26567b  
1  726cce9d9800d792d1e0fbb83923ac4c6be4f668  
2  598d17ec200eeb9e27ae829746ccf47c57761bc1

#### The size of toy_fields and toy_yields tables is same (417) and following columns in both the tables match, 
#### org_id , field_id and year
#### I am assuming they can be merged together to form a big table.

In [118]:
# <aghatpande> on 18-Jan-2018
# Combining Field data and Yield data
fieldAndYieldCombinedDF = pd.merge(toyFieldsDF, toyYieldsDF, left_index=True, right_index=True)
len(fieldAndYieldCombinedDF)
fieldAndYieldCombinedDF.head(2)

417

field_id_x  \
0  cb1e3bd821dd673a0c378b16ba97075919aa993a   
1  58d04a7094c1b08e9856932a02104d2a36b7b8ce   

                                   org_id_x  year_x  field_area  \
0  0966e34316e39217e4f26fbe20fc14e24c26567b    2016  307.185719   
1  726cce9d9800d792d1e0fbb83923ac4c6be4f668    2014  213.680671   

  field_area_unit             crop_type  year_y harvest_product  \
0            acre  CROP.CORN.COMMERCIAL    2016           grain   
1            acre  CROP.CORN.COMMERCIAL    2014           grain   

   total_harvested harvested_unit  planted_area  harvested_area  \
0     53290.628852             bu         40.38           560.0   
1     40073.821554             bu           NaN             NaN   

                                 field_id_y  \
0  cb1e3bd821dd673a0c378b16ba97075919aa993a   
1  58d04a7094c1b08e9856932a02104d2a36b7b8ce   

                                   org_id_y  
0  0966e34316e39217e4f26fbe20fc14e24c26567b  
1  726cce9d9800d792d1e0fbb83923ac4c6be4f668

In [119]:
# <aghatpande> on 18-Jan-2018
# Check if there are any NUlls/Missing values in any columns
pd.isnull(fieldAndYieldCombinedDF).sum() > 0

field_id_x         False
org_id_x           False
year_x             False
field_area         False
field_area_unit    False
crop_type          False
year_y             False
harvest_product    False
total_harvested    False
harvested_unit     False
planted_area        True
harvested_area      True
field_id_y         False
org_id_y           False
dtype: bool

#### Since harvested_area column has NaN/Null values, we cannot calculate yeild per org, field, year. We cannot use field_area because field_area will be different than harvested_area

### Lets try to solve the second expectation - We expect field data to have one acreage measure per field
#### Lets check if each field has one acreage

In [120]:
# <aghatpande> on 18-Jan-2018
# Sort the field_area by Ascending order and check if any one has 0 value
toyFieldsDF.sort_values(by = "field_area", inplace=True)
toyFieldsDF.head(5)
fieldWithZeroAcreageDF = toyFieldsDF[toyFieldsDF.field_area == 0]
print("Number of fields with field_area Zero Acreage are ", len(fieldWithZeroAcreageDF))

field_id  \
407  930a31b273111ac76f4f8785152d93e799150f9c   
235  cda306e104b49f08dafdcc3f0ae8f34a2a265ba0   
389  4d5b90ee18d570ec6b0dc27f64e44ee714b71d72   
95   6a0a29666d8f8a14360dae95a0369dceb6561c6f   
328  4f066ef6d07405fb98f7c15a1d8a080013863a9f   

                                       org_id  year  field_area  \
407  598d17ec200eeb9e27ae829746ccf47c57761bc1  2015    2.839143   
235  fe62dee029474d52346542845674529025594056  2017    7.484157   
389  fe62dee029474d52346542845674529025594056  2016    9.699646   
95   598d17ec200eeb9e27ae829746ccf47c57761bc1  2016   10.549866   
328  106e6791d0f181b391cde800e74fdc14d529dc0a  2014   12.536361   

    field_area_unit  
407            acre  
235            acre  
389            acre  
95             acre  
328            acre

('Number of fields with field_area Zero Acreage are ', 0)


#### Lets take a sample field and see if it has one/same acreage 

In [122]:
# <aghatpande> on 18-Jan-2018
# Lets see a sample field
sampleField = toyFieldsDF[toyFieldsDF.field_id == "17ba0791499db908433b80f37c5fbc89b870084b"]
len(sampleField)
sampleField.sort_values(by="year", inplace=True)
sampleField

21

/Users/test/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy



field_id  \
193  17ba0791499db908433b80f37c5fbc89b870084b   
409  17ba0791499db908433b80f37c5fbc89b870084b   
107  17ba0791499db908433b80f37c5fbc89b870084b   
88   17ba0791499db908433b80f37c5fbc89b870084b   
387  17ba0791499db908433b80f37c5fbc89b870084b   
362  17ba0791499db908433b80f37c5fbc89b870084b   
17   17ba0791499db908433b80f37c5fbc89b870084b   
69   17ba0791499db908433b80f37c5fbc89b870084b   
224  17ba0791499db908433b80f37c5fbc89b870084b   
84   17ba0791499db908433b80f37c5fbc89b870084b   
159  17ba0791499db908433b80f37c5fbc89b870084b   
350  17ba0791499db908433b80f37c5fbc89b870084b   
248  17ba0791499db908433b80f37c5fbc89b870084b   
303  17ba0791499db908433b80f37c5fbc89b870084b   
140  17ba0791499db908433b80f37c5fbc89b870084b   
147  17ba0791499db908433b80f37c5fbc89b870084b   
378  17ba0791499db908433b80f37c5fbc89b870084b   
18   17ba0791499db908433b80f37c5fbc89b870084b   
85   17ba0791499db908433b80f37c5fbc89b870084b   
105  17ba0791499db908433b80f37c5fbc89b870084b   
94   17ba0791499db908433b80f37c5fbc89b870084b   

                                       org_id  year  field_area  \
193  fe62dee029474d52346542845674529025594056  2015  290.735574   
409  fe62dee029474d52346542845674529025594056  2015  545.230446   
107  fe62dee029474d52346542845674529025594056  2015  412.953978   
88   fe62dee029474d52346542845674529025594056  2015  317.400731   
387  fe62dee029474d52346542845674529025594056  2015  287.708372   
362  fe62dee029474d52346542845674529025594056  2015  202.271296   
17   fe62dee029474d52346542845674529025594056  2015  685.826880   
69   fe62dee029474d52346542845674529025594056  2015  135.797506   
224  fe62dee029474d52346542845674529025594056  2015  108.780552   
84   fe62dee029474d52346542845674529025594056  2015   97.541415   
159  fe62dee029474d52346542845674529025594056  2015  159.229421   
350  fe62dee029474d52346542845674529025594056  2016  209.996815   
248  fe62dee029474d52346542845674529025594056  2016  556.489916   
303  fe62dee029474d52346542845674529025594056  2016  337.718651   
140  fe62dee029474d52346542845674529025594056  2016  485.123082   
147  fe62dee029474d52346542845674529025594056  2016   93.481398   
378  fe62dee029474d52346542845674529025594056  2017  239.843347   
18   fe62dee029474d52346542845674529025594056  2017  295.680038   
85   fe62dee029474d52346542845674529025594056  2017  320.127683   
105  fe62dee029474d52346542845674529025594056  2017  385.953665   
94   fe62dee029474d52346542845674529025594056  2017   61.462671   

    field_area_unit  
193            acre  
409            acre  
107            acre  
88             acre  
387            acre  
362            acre  
17             acre  
69             acre  
224            acre  
84             acre  
159            acre  
350            acre  
248            acre  
303            acre  
140            acre  
147            acre  
378            acre  
18             acre  
85             acre  
105            acre  
94             acre

#### As we can see above, each field has different field_area (acreage) per year. This may be because 
1. The organization decides to reserve a particular area within a field for harvesting per year.
2. The organization decides to reserve a different areas for different crops within same year.

#### Lets check if sum of areas (per field) per year is same

In [128]:
# <aghatpande> on 18-Jan-2018
# GroupByField and year to calculate acreage per field per year
fieldGroupByDF = toyFieldsDF.groupby(["field_id", "year"]).agg({'field_area':'sum'})
fieldGroupByDF.head(5)

field_area
field_id                                 year            
00ed44d041d6b0f8a538b2815fdb34b2dcee329c 2015  298.069184
                                         2016  367.176503
                                         2017  548.764229
02dc3fe6f889f3ece8daf456d26464d381bc9e55 2015  194.621312
03074faf1987f5325db3e511880d28a582293c64 2017  190.517286

#### As seen from above result, even the field_areas per year is not same for a given field.

In [48]:
# <aghatpande> on 18-Jan-2018
# Lets explore the toy_orgs data
toyOrgsDF.head(5)

grower                       organization_id  \
0  Cavanaugh Farm  bf08b038-921c-482f-85b7-17c52aa98095   
1     Depner Farm  c0d679bb-50c3-40b2-8a6b-313b6010aae6   
2     Austin Farm  c9a331d4-68fb-4946-9e4e-fdca1eb268b3   
3      Batte Farm  8bbd1e4f-8783-433d-ac74-27e81d67fb43   
4  Sifuentes Farm  35cedd2e-a117-4713-983f-3574fffce422   

                                 new_org_id  
0  fe62dee029474d52346542845674529025594056  
1  106e6791d0f181b391cde800e74fdc14d529dc0a  
2  0966e34316e39217e4f26fbe20fc14e24c26567b  
3  513612757cd6094d98b6a82a01d6f115fe9ae88c  
4  2ec53b20e3a9f6ff73d290ef3e8e8f30a2a30235

In [129]:
# <aghatpande> on 18-Jan-2018
# Lets check if we have any missing data/Nulls
pd.isnull(toyOrgsDF).sum() > 0

grower             False
organization_id    False
new_org_id         False
dtype: bool